File: .orkestrel/campaign/docs-parity/instruments/d7/units/qualifier/d7n-qualifier-converge/cells.mjs
// Compares the key and Kind cells of every guide table row between two files.
import { readFileSync } from 'node:fs'

function splitRow(line) {
	const parts = []
	let cell = ''
	for (let at = 0; at < line.length; at += 1) {
		const char = line[at]
		if (char === '|' && line[at - 1] !== '\\') {
			parts.push(cell.trim())
			cell = ''
			continue
		}
		cell += char
	}
	parts.push(cell.trim())
	return parts.slice(1, -1)
}

function readTables(path) {
	const lines = readFileSync(path, 'utf8').split('\n')
	const rows = new Map()
	for (let at = 0; at < lines.length; at += 1) {
		const line = lines[at]
		if (!line.startsWith('| ')) continue
		if (/^\|\s*-+\s*\|/.test(line)) continue
		const cells = splitRow(line)
		if (cells.length < 2) continue
		if (cells[1] === 'Kind' || cells[0] === 'Type' || cells[0] === 'API' || cells[0] === 'Method') continue
		rows.set(cells[0], cells[1])
	}
	return rows
}

const before = readTables(process.argv[2])
const after = readTables(process.argv[3])
let moved = 0
for (const [key, kind] of before) {
	const now = after.get(key)
	if (now === undefined) { console.log(`missing after: ${key}`); moved += 1; continue }
	if (now !== kind) { console.log(`kind moved: ${key}: ${kind} -> ${now}`); moved += 1 }
}
for (const key of after.keys()) {
	if (!before.has(key)) { console.log(`added: ${key}`); moved += 1 }
}
console.log(`rows before: ${before.size}, rows after: ${after.size}, moved: ${moved}`)

File: .orkestrel/campaign/docs-parity/instruments/d7/units/worker/d7n-worker-converge/cells.py
"""Compares every non-Summary cell of every guide table row against the baseline.

Rows are keyed by their first cell. A pipe preceded by a backslash is not a column
separator, so the split honours the escape a `Shape` or `Signature` cell carries.
"""

import re
import subprocess
import sys

SPLIT = re.compile(r"(?<!\\)\|")


def rows(text):
    found = {}
    for line in text.splitlines():
        stripped = line.strip()
        if not stripped.startswith("|"):
            continue
        cells = [cell.strip() for cell in SPLIT.split(stripped)[1:-1]]
        if not cells or set("".join(cells)) <= {"-", ":"}:
            continue
        found.setdefault(cells[0], []).append(cells)
    return found


before = rows(subprocess.run(["git", "show", "HEAD:guides/worker.md"], capture_output=True, text=True, check=True).stdout)
after = rows(open("guides/worker.md", encoding="utf8").read())

compared = 0
mismatched = []
missing = []
for key, old_list in before.items():
    new_list = after.get(key)
    if new_list is None:
        missing.append(key)
        continue
    for old, new in zip(old_list, new_list):
        compared += 1
        # The final column is the compared Summary cell; every earlier column must travel.
        if old[:-1] != new[: len(old) - 1]:
            mismatched.append((key, old[:-1], new[: len(old) - 1]))

sys.stdout.write(f"rows compared: {compared}\n")
sys.stdout.write(f"rows missing after: {missing}\n")
for key, old, new in mismatched:
    sys.stdout.write(f"MOVED {key}\n  before {old}\n  after  {new}\n")
sys.stdout.write(f"non-Summary cells moved: {len(mismatched)}\n")
